### PR TITLE
Optimize store.set entity validation

### PR DESCRIPTION
**Pull request by leoyvens**
If the entity that comes from the mapping is already valid, avoid going to the DB at all. Also add a metrics section for `as_modifications`, since some DB accesses shift from `store_set` to `as_modifications`. Part of #1381.

PR for grafana https://github.com/graphprotocol/instrumentation/pull/3.

### CODE REVIEW

**Comment by lutter** [comment]
This looks good, but you should also change the lookup that happens in [EntityCache](https://github.com/graphprotocol/graph-node/blob/master/graph/src/components/store.rs#L1210-L1212) from looking up missing entities one by one to using a bulk query (`EntityQuery` with a `EntityFilter::In`) and then put the entities you get back or `None` into `self.current`.

File: runtime/wasm/src/host_exports.rs [hunk 1]
```diff
@@ -142,17 +142,20 @@ impl HostExports {
...
-
-        if self.store.uses_relational_schema(&self.subgraph_id)? {
-            // Validate the changes against the subgraph schema
+        let entity = Entity::from(data);
+        state.entity_cache.set(key.clone(), entity.clone());
+
+        // Validate the changes against the subgraph schema.
+        // If the set of fields we have is already valid, avoid hitting the DB.
+        let schema = self.store.input_schema(&self.subgraph_id)?;
+        if validate_entity(&schema.document, &key, &entity).is_err()
```
Comment (lutter) [inline runtime/wasm/src/host_exports.rs]: You could save a clone of the entity if you did the first `validate_entity` before the `entity_cache.set`.

**Comment by Jannis** [comment]
@leoyvens Were you able to measure the impact of this change compared to before?

File: runtime/wasm/src/host_exports.rs (hunk 1 above)
Comment (leoyvens) [reply runtime/wasm/src/host_exports.rs]: Ah yes I should do that.

**Comment by leoyvens** [comment]
@Jannis I only observed that `store.set` now takes less time than `as_modifications`, as expected. I want to do the bulk loading as well and then I'll measure.

**Comment by Jannis** [comment]
@leoyvens Works for me!

File: runtime/wasm/src/host_exports.rs (hunk 1 above)
Comment (leoyvens) [reply runtime/wasm/src/host_exports.rs]: Done, merging.

**Pull request merged**